### PR TITLE
Spanner custom conversions used in composite key accessor

### LINKED ISSUE #1226: Spanner use custom converters when forming composite key from id properties
Spanner composite keys can only be constructed from set of allowed types. Currently in the few places where this is done there is no conversion done. We should use the same custom converters that are used for the other properties in those cases.

### PR DESCRIPTION
**Pull request by ChengyuanZhao**
Fixes #1226

Utilizes custom converters in the composite key id accessor. 
We had the logic all built, it just needed to be used.

### CODE REVIEW

File: spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/ConverterAwareMappingSpannerEntityWriter.java [hunk 1]
```diff
@@ -49,7 +49,7 @@
  * @since 1.1
  */
 public class ConverterAwareMappingSpannerEntityWriter implements SpannerEntityWriter {
-	private static Set<Class> SPANNER_KEY_COMPATIBLE_TYPES = ImmutableSet
+	private final static Set<Class> SPANNER_KEY_COMPATIBLE_TYPES = ImmutableSet
```
[the remark this answers is not in the document]
Comment (ChengyuanZhao) [reply spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/ConverterAwareMappingSpannerEntityWriter.java]: you gotta be kidding me lol. 
will fix.

File: spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/SpannerKeyWriter.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	 * @return the Cloud Spanner key.
+	 */
+	Key writeToKey(Object key);
+
+	/**
+	 * Get the SpannerWriteConverter used to convert types into Cloud Spanner compatible
+	 * types.
+	 * @return a SpannerWriteConverter
+	 */
+	SpannerWriteConverter getSpannerWriteConverter();
```
Comment (meltsufin) [inline spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/SpannerKeyWriter.java]: What does this have to do with keys?

File: spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/SpannerKeyWriter.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ */
+public interface SpannerKeyWriter {
+
+	/**
+	 * Writes a given object to a Cloud Spanner key.
+	 * @param key the object containing the key values. This can already be a Cloud Spanner
+	 *     key, a single key component, or an array of key components.
+	 * @return the Cloud Spanner key.
+	 */
+	Key writeToKey(Object key);
```
Comment (meltsufin) [inline spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/SpannerKeyWriter.java]: It seems more like a `convertToKey` method.

File: spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/SpannerEntityWriter.java [hunk 4]
```diff
@@ -38,6 +36,4 @@
 	 * written.
 	 */
 	void write(Object source, MultipleValueBinder sink, Set<String> includeColumns);
-
-	Key writeToKey(Object key);
```
Comment (meltsufin) [inline spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/SpannerEntityWriter.java]: What's the benefit of pulling this out into the `SpannerKeyWriter` interface?

File: spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/SpannerKeyWriter.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import com.google.cloud.spanner.Key;
+
+/**
+ * An interface that can create Cloud Spanner keys and utlizes conversion.
+ *
+ * @author Chengyuan Zhao
+ *
+ * @since 1.1
+ */
+public interface SpannerKeyWriter {
```
Comment (meltsufin) [inline spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/SpannerKeyWriter.java]: I'm not sure why this interface is useful vs. keeping these methods in `SpannerEntityWriter`.

Comment (ChengyuanZhao) [reply spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/SpannerKeyWriter.java]: The reason is this new one is what's added to the constructor of SpannerPersistentEntityImpl because it is needed to convert  the properties that comprise the composite key.

File: spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/SpannerEntityWriter.java (hunk 4 above)
Comment (ChengyuanZhao) [reply spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/SpannerEntityWriter.java]: The reason is this new one is what's added to the constructor of SpannerPersistentEntityImpl because it is needed to convert  the properties that comprise the composite key. We don't need the write-to-entity functionality, but only the write-to-key functionality.

File: spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/SpannerKeyWriter.java (hunk 3 above)
Comment (ChengyuanZhao) [reply spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/SpannerKeyWriter.java]: will do

File: spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/SpannerKeyWriter.java (hunk 2 above)
Comment (ChengyuanZhao) [reply spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/SpannerKeyWriter.java]: This converter is needed to convert the individual properties that comprise the composite key.

File: spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/SpannerKeyWriter.java (hunk 5 above)
Comment (meltsufin) [reply spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/SpannerKeyWriter.java]: Why would giving `SpannerEntityWriter` to the `SpannerPersistentEntityImpl` be so bad? You never implement  just `SpannerKeyWriter`, and it looks like an internal class.

Comment (ChengyuanZhao) [reply spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/SpannerKeyWriter.java]: To be honest I personally don't think it's bad, but I did it because I thought if I gave it a whole entity-converter interface to something that only uses the key-converter functions would offend somebody with more tender java sensibilities. I can get rid of the interface.

Comment (ChengyuanZhao) [reply spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/SpannerKeyWriter.java]: done
